main/autogen/autogen.cc

```diff
@@ -674,4 +676,331 @@ void ParsedFile::classlist(core::Context ctx, vector<string> &out) {
     }
 }
 
+NamedDefinition ParsedFile::toNamed(core::Context ctx, DefinitionRef def) {
+    auto names = showFullName(ctx, def);
+    vector<core::NameRef> parentName;
+    if (def.data(*this).parent_ref.exists()) {
+        auto parentRef = def.data(*this).parent_ref.data(*this);
+        if (!parentRef.resolved.empty()) {
+            parentName = parentRef.resolved;
+        } else {
+            parentName = parentRef.name;
+        }
+    }
+    return {def.data(*this), names, parentName, requires, tree.file};
+}
+
+void ParsedFile::addDefinitions(core::Context ctx, const AutoloaderConfig &alConfig, DefTree &root) {
+    if (!alConfig.includePath(path)) {
+        return;
+    }
+    for (auto &def : defs) {
+        if (def.id.id() == 0) {
+            continue;
+        }
+        NamedDefinition namedDef = toNamed(ctx, def.id);
+        root.addDef(ctx, alConfig, move(namedDef));
+    }
+}
+
+void DefTree::addDef(core::Context ctx, const AutoloaderConfig &alCfg, NamedDefinition ndef) {
+    if (!alCfg.include(ndef)) {
+        return;
+    }
+    auto *node = this;
```

Comment:
if this wasn't a member of `DefTree`, this function would be easier to express as it could operate on `unique_ptr<DefTree> &`